### ISSUE TITLE
Research on open source multiUxVs teaming modules. 

### ISSUE THREAD

**Issue by okyspace** [DinoHub/devmgmt-monkeybotics#99]

**Comment by jonyktan** [comment]
Doc here: https://docs.google.com/spreadsheets/d/1E17JULxcLi1S6DPnhjbzPzdRZzL1hJ59/edit?usp=sharing&ouid=114674236949679804376&rtpof=true&sd=true

Started to create headers for "features".

**Comment by okyspace** [comment]
Read up VIS4ROB.

**Comment by jonyktan** [comment]
## **VIS4ROB**
Read up on VIS4ROB [here](https://www.research-collection.ethz.ch/bitstream/handle/20.500.11850/441280/IROS_2020_Multi_Robot_compressed_v4.pdf?sequence=1&isAllowed=y). 

1. The work proposes a system architecture that comprises a server and a team of agents (small aircraft).
1. Computationally demanding tasks (global dense mapping, global path planning) are outsourced to the central server.
1. Onboard computation on each agent is limited to local pose estimation using VIO, local path planning for obstacle avoidance, to keep within the typical design constraints associated with small UAVs (i.e. low payload capacity).
1. Work is probably not so relevant in our case as it requires reliable communication between agents and server.

### General Outline
**Global Map**

1. VIO onboard each agent estimates agent pose.
2. Central server stores agent poses in a pose-graph. Pose-graph is optimized locally for each agent and globally for multiple agents.
3. Separate threads run independent local optimization for each agent using N most recent keyframes.
4. Global optimization is performed upon detection of a loop closure, on agent(s) with trajector(ies) associated with the loop closure.
5. Each agent transmits point clouds that are compressed by fusing point clouds across frames using voxel filtering.
6. Server creates global dense map from the agent point clouds and optimized poses.

**Global path planner** 

1. Each agent must be assigned an area to cover at the start of each mission. In the form of sequence of GPS waypoints. Unreachable waypoints are discarded by the planner.
2. Paths between waypoints are computed using standard version of RRT* from OMPL library.
3. Collisions between agents are avoided using hierarchy in planning - Agents are arbitrarily assigned hierarchy; higher-priority agents have paths planned first within known free space; lower-priority agents have paths planned later, with their free space diminished by the paths of higher-priority agents.
4. Paths are checked for collision with the (updated) global map at a fixed rate.
5. Local path planning and obstacle avoidance is run onboard each agent by optimizing local trajectory with a cost function.
6. Upon completion of exploration, agents return to respective home position by planning only in explored and known space.

### Evaluation
1. System evaluated on Machine Hall sequences of EuRoC dataset, with Gaussian noise injected on GT position measurements to simulate GPS.
1. System also evaluated on photorealistic environment (120m x 120m chemical plant) in Gazebo using RotorS framework.

**Comment by timothymah** [comment]
Repos to look into:
http://wiki.ros.org/multirobot_map_merge#Acknowledgements

https://github.com/Yaaximus/multi-agent-exploration

https://github.com/JDaniel41/Autonomous-Ci-Multi-Agent-Mapping

**Comment by jonyktan** [comment]
End of sprint - task not complete. Estimate 2hrs/algo to read up on.

**Comment by jonyktan** [comment]
## **ZJU EGO-Swarm** [WIP]
Read up on EGO-Swarm [here](https://arxiv.org/pdf/2011.04183.pdf). 

1. Extension of previous work **ESDF-free Gradient-based Local Planner** (EGO-Planner) - onboard local planning of single quadrotor in unknown environments.

**Comment by jonyktan** [comment]
## **Decentralized Strategy for Cooperative Multi-Robot Exploration and Mapping**
Read the paper [here](https://www.sciencedirect.com/science/article/pii/S2405896320333760). 

**tl;dr - Possibly useful. Can learn from their architecture that uses event-based updates between robots and a decentralised optimisation on each robot to select the next frontier for exploration.
